Simplify ECR Automation Workflows

cabutlermit opened the pull request:
Why these changes are being introduced:
The carbon build process required a few extra steps before `docker build` would run correctly. For automated deploys to dev & stage and promotes to prod to work, the shared workflows needed to be able to handle those additional steps. Additionally, there were duplicate workflows for Fargate deploys and Lambda deploys that should be consolidated for simpler (and DRYer) code. Finally, this repo should have a PR template like all our other repos.

How this addresses that need:
* Create a single new ecr-shared-deploy-dev workflow file that will handle both Fargate containers and Lambda containers/functions
* Create a single new ecr-shared-deploy-stage workflow file that will handle both Fargate containers and Lambda containers/functions
* Create a single new ecr-shared-promote-prod workflow file that will handle both Fargate containers and Lambda containers/functions
* Add checks for `.python-version` and `.ruby-version` files in the repo (in the new dev and stage workflows)
* Clean up the setup python step using actions/setup-python only if the `.python-version` file exists in the root of the repo (in the new dev and stage workflows)
* Add a setup ruby step using ruby/setup-ruby only if the `.ruby-version` file exists in the root of the repo (in the new dev and stage workflows)
* Clean up the "pre-build" step so that it is universal (only depends on the caller workflow specifying the commands)
* Switch to using $GITHUB_ENV from the deprecated set-output method
* Update the aws-credential action to use the Node.js 16 version
* Update in-line comments
* Update README
* Add a PR template file

Side effects of this change:
None. We leave the old workflows in place for now, until we can update all the dependent repos.

cabutlermit committed:
Simplify ECR Deploys

Why these changes are being introduced:
The carbon build process required a few extra steps before
`docker build` would run correctly. For automated deploys to dev & stage
and promotes to prod to work, the shared workflows needed to be able
to handle those additional steps. Additionally, there were duplicate
workflows for Fargate deploys and Lambda deploys that should be
consolidated for simpler (and DRYer) code. Finally, this repo should
have a PR template like all our other repos.

How this addresses that need:
* Create a single new ecr-shared-deploy-dev workflow file that will
handle both Fargate containers and Lambda containers/functions
* Create a single new ecr-shared-deploy-stage workflow file that will
handle both Fargate containers and Lambda containers/functions
* Create a single new ecr-shared-promote-prod workflow file that will
handle both Fargate containers and Lambda containers/functions
* Add checks for `.python-version` and `.ruby-version` files in the
repo (in the new dev and stage workflows)
* Clean up the setup python step using actions/setup-python only if
the `.python-version` file exists in the root of the repo (in the new
dev and stage workflows)
* Add a setup ruby step using ruby/setup-ruby only if the
`.ruby-version` file exists in the root of the repo (in the new dev and
stage workflows)
* Clean up the "pre-build" step so that it is universal (only depends
on the caller workflow specifying the commands)
* Switch to using $GITHUB_ENV from the deprecated set-output method
* Update the aws-credential action to use the Node.js 16 version
* Update in-line comments
* Update README
* Add a PR template file

Side effects of this change:
None. We leave the old workflows in place for now, until we can
update all the dependent repos.

cabutlermit commented:
@hakbailey , @JPrevost , and @zotoMIT 
I think that this needs reviews from all three of you for various parts.
1) Helen: verification that the Python steps in the dev and stage workflows make sense. This needs to work for at least carbon, and not break any other Python-based containerized apps.
2) Jeremy: verification that the Ruby steps in the dev and stage workflows make sense (we've already discussed this a bit on the @engineering thread)
3) Aaron: verification that the docker build and push commands don't break anything for dev, stage, and the promote-to-prod.

Thanks!

zotoMIT reviewed (approved):
Looks good to me.

hakbailey commented on .github/workflows/ecr-shared-deploy-dev.yml:
```diff
@@ -0,0 +1,123 @@
+          echo "tag_sha=${TAG_SHA}" >> $GITHUB_ENV
+          echo "tag_pr=${TAG_PR}" >> $GITHUB_ENV
+
+      - name: Get Python
```
I think we should be able to remove these Get Python and Setup Python steps. If those extra steps are needed pre-Docker build for a given repo, they should be passed into PREBUILD. Generally they should not be necessary.

cabutlermit replied:
I'm going to leave the "Get Python" step for now but remove the "Setup Python" step. That way, everything is ready in case there needs to be a pre-docker command passed into PREBUILD.

cabutlermit replied:
Per Helen's approval, this can be resolved.

cabutlermit committed:
Clean Up Python Setup

* Remove the "Setup Python" step in the dev and stage workflows

This leaves the "Get Python" in place so that the environment is ready
for any Python related commands that the developer passes into PREBUILD.

JPrevost commented on .github/workflows/ecr-shared-deploy-dev.yml:
```diff
@@ -0,0 +1,117 @@
+        run: |
+          if [[ -f ".python-version" ]]; then
+            echo "python_repo=1" >> $GITHUB_ENV
+          else [[ -f ".ruby-version" ]]; 
```
This syntax looks weird. This doesn't mean it is wrong by any means, but it _feels_ like if/else logic but it should function as an if/else_if. Are we confident this works as intended? Even if it does, I'm curious if separating this logic into separate Check for python with an if, then a complete separate check for ruby might be much clearer with a small downside potential of having a repo where we oddly grab both (but maybe that's a feature?).

cabutlermit replied:
It does look a little weird, but in my head I was thinking "it's either a Python repo or a Ruby repo or a simple Docker repo", so the if/else instead of if/elif felt okay. So far, in my testing, it has worked as expected (the [docker-matomo](https://github.com/MITLibraries/docker-matomo) repo which has neither a .`ruby-version` nor a `.python-version` file doesn't attempt to prep either Python or Ruby).

That said, I'm totally comfortable switching this to 

```bash
if [[ -f ".python-version" ]]; then
    echo "python_repo=1" >> $GITHUB_ENV
fi
if [[ -f ".ruby-version" ]]; then
    echo "ruby_repo=1" >> $GITHUB_ENV
fi
```

JPrevost replied:
I think that's clearer to me what is going on and I'd prefer it but this is definitely not a blocker at all if the other syntax is more correct in any way (i.e. if it works, it works 👍🏻 )

JPrevost commented on .github/workflows/ecr-shared-deploy-dev.yml:
```diff
@@ -0,0 +1,117 @@
+      - name: Get Python
+        # Run the Python setup only when there is a .python-version file in the root
+        # of the repository. Otherwise, this step is skipped.
+        if: ${{ env.python_repo == 1 }}
```
Would it be possible to remove the previous Ruby/Python check section and just add a condition like ```if [[ -f ".python-version" ]]` here rather than relying that check previously setting a value in the .env? If so, I think that simplifies this a lot by having all the python logic in one place rather than splitting it around the credentials and tags bits that are complete unrelated.

cabutlermit replied:
Possibly. In my testing of the `if:` line in GitHub Actions, it didn't seem work when I tried to check for the existence of a file, so I switched to setting an environment variable instead. I don't know why it didn't work with a direct test for the file -- but I can poke at this a little more.

JPrevost replied:
Cool. If you already went down this path and it didn't work it's totally fine with the env solution.

JPrevost commented on .github/workflows/ecr-shared-deploy-dev.yml:
```diff
@@ -0,0 +1,117 @@
+        # root of the repository. Otherwise, this step is skipped.
+        if: ${{ env.ruby_repo == 1 }}
+        uses: ruby/setup-ruby@v1
+        with:
```
Previously I had stated I didn't think we needed this but it wouldn't cause harm, but I know think we actually must remove this bundle stuff to avoid build errors later and just use in container building of the bundles.

We _often_ have to add additional libraries to the containers to allow certain gems to build so this is very likely to fail if we don't add those system libraries to our build env which it really shouldn't need.

cabutlermit replied:
I'm totally happy to remove the bundle lines. I think I've stated before that I really have no idea how Ruby works and this was just a copy/paste from the Ruby CI workflows. 😁 

Mostly, the intent of the "Get Ruby" step in the workflow is to make sure that the environment in the runner is ready for any "prebuild" commands that developer needs to run before the `docker build` command. This is necessary for some of our older Python repos (see Carbon and the `make wheel` command which depends on the right version of Python in the runner). But it might not be the case for any of our Ruby repos.

So, we might just need

```yaml
      - name: Get Ruby
        # Run the Ruby setup setup only when there is a .ruby-version file in the 
        # root of the repository. Otherwise, this step is skipped.
        if: ${{ env.ruby_repo == 1 }}
        uses: ruby/setup-ruby@v1
```

JPrevost replied:
Yeah that looks good, thanks.

JPrevost commented on .github/workflows/ecr-shared-deploy-stage.yml:
```diff
@@ -0,0 +1,95 @@
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v1-node16
+        with:
+          role-to-assume: arn:aws:iam::${{ secrets.AWS_ACCT_STAGE }}:role/${{ inputs.GHA_ROLE }}
```
Is there any difference between the dev/stage workflows other than which secret they pull?

cabutlermit replied:
Slightly. The dev workflow needs to do some extra work around tagging in order to make the takes that appear in AWS actually match the commit hashes that are visible in GitHub (see lines 60-76). This isn't necessary in the stage workflow because it's triggered by a push to `main` (where the dev workflow is triggered by a PR).

But, with a little extra work, it's likely pretty straightforward to pull these two shared workflows into one.

JPrevost replied:
Gotcha. They looked so similar I get nervous about unintentional drift. I think that can be resolved in the future though and trying to merge them would likely be premature optimization (i.e. let's see how this _actually_ works over time before we try to make it perfect).

JPrevost reviewed (changes requested):
I think we need to remove the ruby bundling and I'd like confirmation on the syntax of the conditional logic.

Otherwise, this is quite interesting and looks good.

cabutlermit committed:
Address PR feedback re: Ruby

Why these changes are being introduced:
A few concerns were raised concerning the Ruby setup step and some of
the logic in the workflow.

How this addresses that need:
* Simplify the if..then checks for Python and Ruby in the dev and stage
workflows
* Remove unnecessary steps from the Ruby setup step in the dev and
stage workflows
* Minor changes to in-line comments in the dev and stage workflows

Side effects of this change:
None.

cabutlermit commented:
@hakbailey and @JPrevost I have pushed a couple of commits to address the concerns you raised. To be honest, I'm not sure of the protocol for resolving conversations/requested changes, but it feels best to me if you can re-review with my new changes and resolve the conversations if you feel that I've addressed them appropriately.

hakbailey reviewed (approved):
👍
